### ISSUE TITLE
Installation issue: py-numpy

### ISSUE THREAD

**Issue by wspear** [LLNL/spack#45426]
### Steps to reproduce the issue

`spack install py-numpy@2.0.1`

I first noticed this error installing py-scipy@1.13.1. I'm not sure why/how installing py-scipy references the py-numpy's package.py file.

<details><summary>spack spec -I py-numpy@2.0.1</summary>

<pre>
Input spec
--------------------------------
 -   py-numpy@2.0.1

Concretized
--------------------------------
 -   py-numpy@2.0.1%gcc@11.4.0 build_system=python_pip patches=873745d arch=linux-ubuntu22.04-westmere
[+]      ^gcc-runtime@11.4.0%gcc@11.4.0 build_system=generic arch=linux-ubuntu22.04-westmere
[e]      ^glibc@2.35%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-westmere
[e]      ^openblas@0.3.26%gcc@11.4.0~bignuma~consistent_fpcsr+dynamic_dispatch+fortran~ilp64+locking+pic+shared build_system=makefile symbol_suffix=none threads=none arch=linux-ubuntu22.04-westmere
[+]      ^py-cython@3.0.10%gcc@11.4.0 build_system=python_pip arch=linux-ubuntu22.04-westmere
[+]          ^py-setuptools@69.2.0%gcc@11.4.0 build_system=generic arch=linux-ubuntu22.04-westmere
[+]      ^py-meson-python@0.16.0%gcc@11.4.0 build_system=python_pip arch=linux-ubuntu22.04-westmere
[+]          ^meson@1.3.2%gcc@11.4.0 build_system=python_pip patches=0f0b1bd arch=linux-ubuntu22.04-westmere
[+]              ^ninja@1.12.0%gcc@11.4.0+re2c build_system=generic arch=linux-ubuntu22.04-westmere
[+]                  ^python@3.9.19%gcc@11.4.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=0d98e93,4c24573,ebdca64,f2fd060 arch=linux-ubuntu22.04-westmere
[+]                  ^re2c@3.0%gcc@11.4.0 build_system=generic arch=linux-ubuntu22.04-westmere
[+]          ^py-packaging@23.1%gcc@11.4.0 build_system=python_pip arch=linux-ubuntu22.04-westmere
[+]              ^py-flit-core@3.9.0%gcc@11.4.0 build_system=python_pip arch=linux-ubuntu22.04-westmere
[+]          ^py-pyproject-metadata@0.7.1%gcc@11.4.0 build_system=python_pip arch=linux-ubuntu22.04-westmere
[+]      ^py-pip@23.1.2%gcc@11.4.0 build_system=generic arch=linux-ubuntu22.04-westmere
[+]      ^py-wheel@0.41.2%gcc@11.4.0 build_system=generic arch=linux-ubuntu22.04-westmere
[+]      ^python@3.11.9%gcc@11.4.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=13fa8bf,b0615b2,ebdca64,f2fd060 arch=linux-ubuntu22.04-westmere
[+]          ^bzip2@1.0.8%gcc@11.4.0~debug~pic+shared build_system=generic arch=linux-ubuntu22.04-westmere
[+]              ^diffutils@3.10%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-westmere
[+]          ^expat@2.6.2%gcc@11.4.0+libbsd build_system=autotools arch=linux-ubuntu22.04-westmere
[+]              ^libbsd@0.12.2%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-westmere
[+]                  ^libmd@1.0.4%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-westmere
[+]          ^gdbm@1.23%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-westmere
[+]          ^gettext@0.22.5%gcc@11.4.0+bzip2+curses+git~libunistring+libxml2+pic+shared+tar+xz build_system=autotools arch=linux-ubuntu22.04-westmere
[+]              ^libiconv@1.17%gcc@11.4.0 build_system=autotools libs=shared,static arch=linux-ubuntu22.04-westmere
[+]              ^libxml2@2.10.3%gcc@11.4.0+pic~python+shared build_system=autotools arch=linux-ubuntu22.04-westmere
[+]              ^tar@1.34%gcc@11.4.0 build_system=autotools zip=pigz arch=linux-ubuntu22.04-westmere
[+]                  ^pigz@2.8%gcc@11.4.0 build_system=makefile arch=linux-ubuntu22.04-westmere
[+]                  ^zstd@1.5.6%gcc@11.4.0+programs build_system=makefile compression=none libs=shared,static arch=linux-ubuntu22.04-westmere
[+]          ^gmake@4.4.1%gcc@11.4.0~guile build_system=generic arch=linux-ubuntu22.04-westmere
[+]          ^libffi@3.4.6%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-westmere
[+]          ^libxcrypt@4.4.35%gcc@11.4.0~obsolete_api build_system=autotools patches=4885da3 arch=linux-ubuntu22.04-westmere
[+]              ^perl@5.38.2%gcc@11.4.0+cpanm+opcode+open+shared+threads build_system=generic patches=714e4d1 arch=linux-ubuntu22.04-westmere
[+]                  ^berkeley-db@18.1.40%gcc@11.4.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-ubuntu22.04-westmere
[+]          ^ncurses@6.5%gcc@11.4.0~symlinks+termlib abi=none build_system=autotools patches=7a351bc arch=linux-ubuntu22.04-westmere
[+]          ^openssl@3.3.1%gcc@11.4.0~docs+shared build_system=generic certs=mozilla arch=linux-ubuntu22.04-westmere
[+]              ^ca-certificates-mozilla@2023-05-30%gcc@11.4.0 build_system=generic arch=linux-ubuntu22.04-westmere
[+]          ^pkgconf@2.2.0%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-westmere
[+]          ^readline@8.2%gcc@11.4.0 build_system=autotools patches=bbf97f1 arch=linux-ubuntu22.04-westmere
[+]          ^sqlite@3.46.0%gcc@11.4.0+column_metadata+dynamic_extensions+fts~functions+rtree build_system=autotools arch=linux-ubuntu22.04-westmere
[+]          ^util-linux-uuid@2.40.2%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-westmere
[+]          ^xz@5.4.6%gcc@11.4.0~pic build_system=autotools libs=shared,static arch=linux-ubuntu22.04-westmere
[+]          ^zlib-ng@2.2.1%gcc@11.4.0+compat+new_strategies+opt+pic+shared build_system=autotools arch=linux-ubuntu22.04-westmere
[+]      ^python-venv@1.0%gcc@11.4.0 build_system=generic arch=linux-ubuntu22.04-westmere

</pre></details>

### Error message

<details><summary>Error message</summary>

<pre>
==> Installing py-numpy-2.0.1-xkruhx6soqouilzmymqxkvhavt2ckkev [35/35]
==> No binary for py-numpy-2.0.1-xkruhx6soqouilzmymqxkvhavt2ckkev found: installing from source
==> Using cached archive: /home/wspear/bin/SPACK/spack/var/spack/cache/_source-cache/archive/48/485b87235796410c3519a699cfe1faab097e509e90ebb05dcd098db2ae87e7b3.tar.gz
==> Applied patch /home/wspear/bin/SPACK/spack/var/spack/repos/builtin/packages/py-numpy/check_executables.patch
==> py-numpy: Executing phase: 'install'
==> Error: IndexError: list index out of range

/home/wspear/bin/SPACK/spack/var/spack/repos/builtin/packages/py-numpy/package.py:230, in blas_lapack_pkg_config:
        227            The names of the blas and lapack libs that pkg-config should search for.
        228        """
        229        spec = self.spec
  >>    230        blas = spec["blas"].libs.names[0]
        231        lapack = spec["lapack"].libs.names[0]
        232
        233        if spec["blas"].name in ["intel-mkl", "intel-parallel-studio", "intel-oneapi-mkl"]:

See build log for details:
  /tmp/wspear/spack-stage/spack-stage-py-numpy-2.0.1-xkruhx6soqouilzmymqxkvhavt2ckkev/spack-build-out.txt

</pre></details>

### Information on your system

* **Spack:** 0.23.0.dev0 (e5364ea832abb7ed79f4cbdc7addf8afe6bb353f)
* **Python:** 3.10.12
* **Platform:** linux-ubuntu22.04-westmere
* **Concretizer:** clingo

### Additional information

@adamjstewart @rgommers 

Logs: http://yu.nic.uoregon.edu/~wspear/py-numpy-error/

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by wspear** [comment]
To clarify, the error also showed up during py-scipy's install step, but somehow occurred in py-numpy's package.

```
==> Installing py-scipy-1.13.1-k25vrarqupynhkmkxndmvtsxoovzgo6m [84/88]
==> No binary for py-scipy-1.13.1-k25vrarqupynhkmkxndmvtsxoovzgo6m found: installing from source
==> Fetching https://files.pythonhosted.org/packages/source/s/scipy/scipy-1.13.1.tar.gz
==> No patches needed for py-scipy
==> py-scipy: Executing phase: 'install'
==> Error: IndexError: list index out of range

/home/wspear/bin/SPACK/spack/var/spack/repos/builtin/packages/py-numpy/package.py:230, in blas_lapack_pkg_config:
        227            The names of the blas and lapack libs that pkg-config should search for.
        228        """
        229        spec = self.spec
  >>    230        blas = spec["blas"].libs.names[0]
        231        lapack = spec["lapack"].libs.names[0]
        232
        233        if spec["blas"].name in ["intel-mkl", "intel-parallel-studio", "intel-oneapi-mkl"]:
```

**Comment by adamjstewart** [comment]
How did you encounter an error when building scipy if the numpy build is also broken? You can't install scipy without first successfully installing numpy.

**Comment by adamjstewart** [comment]
It looks like you're using OpenBLAS. Can you share the contents of the lib directory of OpenBLAS?

**Comment by adamjstewart** [comment]
> I'm not sure why/how installing py-scipy references the py-numpy's package.py file.

why: we want to build SciPy with the same BLAS/LAPACK flags as NumPy
how: https://github.com/spack/spack/blob/v0.22.1/var/spack/repos/builtin/packages/py-scipy/package.py#L217

This is expected, the failure for Spack to find OpenBLAS's libs is not.

**Comment by wspear** [comment]
Yep, user error. I had to nuke my previous spack install but my global packages.yaml was still specifying an old openblas (provided by that old spack) as an external that I needed pinned for some reason lost to history. It would be nice if missing externals were caught further up the chain.

I'm not sure why/how I got to build scipy with numpy broken. It happened in the course of building py-deephyper.
